### ISSUE TITLE
Prepare for release of Plone 5.2a1 with python 3 support

### ISSUE THREAD

**Issue by pbauer** [plone/Products.CMFPlone#2592]
Here we discuss and collect all tasks that need to be done before a 5.2a1 release.

* [x] Fix all tests in python 3.6  (https://jenkins.plone.org/view/PLIPs/job/plip-py3)
* [x] Fix all tests in python 3.7 (https://jenkins.plone.org/view/PLIPs/job/plip-py3.7)
* [x] Fix all tests in python 2.7 (https://jenkins.plone.org/view/PLIPs/job/py3_on_py2)
* [x] Activate pull-request jobs for 5.2 that use py36 and py37
* [x] Reconfigure coredev buildout. See #2593
* [x] Merge all unmerged branches from py3.cfg and py3_on_py3.cfg.
* [x] Update to Zope 4.0b7 and the versions from that. See #2590

Please add additional tasks that are a hard requirement for a release of Plone 5.2a1.

**Comment by pbauer** [comment]
The failing tests in plone.restapi are all from `TestBatchingArchetypes` and they fail together with the robot-tests in `plone.app.widgets`. Even when I strip everything from the tests and the testsetup of the robot-tests in plone.app.widgets it fails. 
To reproduce: 
```
./bin/test --all -s plone.app.widgets -s plone.restapi -t test_datetime_widget.robot -t TestBatchingArchetypes
```

**Comment by Rotonen** [comment]
The filedescriptor leaks should also be tackled for testing stability. I'm having a feeling currently the Robot tests can get stuck when hitting an ulimit and the test setup machinery not handling that scenario.

**Comment by pbauer** [comment]
I skipped the tests in `TestBatchingArchetypes` but then next tests `TestATContentPatch` failed. It seems the first tests in `plone.restapi.testing.PloneRestApiATLayer:Functional` fails but consecutive tests do not fail. I tried to add a dummy Test-Class that is run before the others in the hope that the teardown of that would fix consecutive TestCases but that was no successfull. I'm currently at the end of my wits here.

**Comment by Rotonen** [comment]
Run those quick runs with `--shuffle` and `-vvv` to see if the ordering between failures and successes pinpoints the correct leaking test, testclass or layer execution and/or setup/teardown out for you.

**Comment by pbauer** [comment]
For the failing tests `plone.rest.events.mark_as_api_request` is never hit 😟

**Comment by pbauer** [comment]
Could I buy me a clone of @davisagli for all the times that I run into these soul-crunching issues?

**Comment by pbauer** [comment]
I added a crazy workaround for the issue in plone.restapi: https://github.com/plone/plone.restapi/commit/23fad9ad12d11c00c3c524f1ab032b0c5bb54cb3

**Comment by pbauer** [comment]
And... we are green!
![green](https://user-images.githubusercontent.com/453208/46881843-f818a080-ce4c-11e8-86e9-608ebbf070a9.png)

**Comment by thet** [comment]
wohoo!

**Comment by pgrunewald** [comment]
Great news!

**Comment by davisagli** [comment]
Oh dear -- I'll try to find a chance to look at the plone.restapi failure soon. Glad it's green though!

**Comment by tisto** [comment]
@pbauer if p.a.widgets tests bleed into p.restapi tests why don't we just move the package to the test-isolation issues group in buildout? If we start to write that kind of work-around tests we just make things worse IMHO.

**Comment by pbauer** [comment]
@tisto the whole alltests-thing will be gone and the test-isolation group as well. We're running tests with `./bin/test --all` because we fixed almost all test-isolation issues.

**Comment by tisto** [comment]
@pbauer wow, ok. That's great news! Never thought that the work I started many years ago actually would come to an end at some point. :)

Though unfortunately being able to run "bin/test --all" just means that we currently do not see test isolation issues in the specific order we run the tests. The test isolation issues are still present.

If we now start to fix the problems of one package in another package, we risk that their test fixtures are tightly coupled to each other and we might even not be able to run their tests successfully individually any longer. Since Jenkins only runs "bin/test --all", we might not even detect that package tests don't pass individually any longer. 

Therefore keeping alltests might be a better option than writing tightly coupled test fixtures. I don't have a single simple answer to this complex problem. Maybe the way zope.testrunner optimizes the test fixture is just wrong in the first place...

**Comment by lukasgraf** [comment]
I've looked into the `plone.restapi` failure last night.

The failure in `plone.restapi` happens because the subscriber registration of the [`mark_as_api_request` for `IPubStart`](https://github.com/plone/plone.rest/blob/b43bf5266876b41256772a0ab2627e6a412588b5/src/plone/rest/configure.zcml#L8-L11) seems to vanish when a robotframework test like the one from `p.a.widgets` runs before it.

This is the difference I'm seeing (all from inside a waitress request handler thread):

---

**Passing test** (just running the `plone.restapi` test alone):

```python
(Pdb) sm = getSiteManager()
(Pdb) print list(sm.registeredHandlers())
[HandlerRegistration(<BaseGlobalComponents test-stack-3>, [IPubStart], u'', mark_as_api_request, File ".../plone/rest/configure.zcml", line 8.2-11.6)]
```

=> Site manager contains the event subscriber registration. Good.

**Failing test** (first the robot test from `p.a.widgets`, then the `plone.restapi` test):

```python
(Pdb) sm = getSiteManager()
(Pdb) print list(sm.registeredHandlers())
[]
```

=> Registration is missing.

---

When I further inspect which exact stacked component registry instance we're working with, I see something strange:

**Passing test:**

```python
(Pdb) sm
<BaseGlobalComponents test-stack-3>
(Pdb) from plone.testing import zca
(Pdb) zca._REGISTRIES
[<BaseGlobalComponents base>, <BaseGlobalComponents test-stack-1>, <BaseGlobalComponents test-stack-2>, <BaseGlobalComponents test-stack-3>]
(Pdb) zca._REGISTRIES[-1] is sm
True
(Pdb) id(zca._REGISTRIES[-1]), id(sm)
(4586995088, 4586995088)
```

=> The current site manager is the last one from `plone.testing.zca`'s component registry stack. Good.

**Failing test:**

```python
(Pdb) sm
<BaseGlobalComponents test-stack-3>
(Pdb) from plone.testing import zca
(Pdb) zca._REGISTRIES
[<BaseGlobalComponents base>, <BaseGlobalComponents test-stack-1>, <BaseGlobalComponents test-stack-2>, <BaseGlobalComponents test-stack-3>]
(Pdb) zca._REGISTRIES[-1] is sm
False
(Pdb) id(zca._REGISTRIES[-1]), id(sm)
(4652502480, 4583929552)
```

=> The current site manager is a different one than the last one from `plone.testing.zca`'s stack, *yet has the same name (`test-stack-3`) / stack depth number*.

---

If we actually poke the last component registry from zca's stack (during the failing test), that's where our subscriber registration would be:

**Failing test:**

```python
(Pdb) print list(zca._REGISTRIES[-1].registeredHandlers())
[HandlerRegistration(<BaseGlobalComponents test-stack-3>, [IPubStart], u'', mark_as_api_request, File ".../plone/rest/configure.zcml", line 8.2-11.6)]
```

This leads be to believe that there's an isolation issue around the component registry with the [`WSGIServer` layer](https://github.com/plone/plone.testing/blob/py3_disallow_commits/src/plone/testing/zope.py#L886-L959).

My hypothesis is this:

The reference to the current site manager is stored on the [`SiteInfo` object in `zope.component.hooks`](https://github.com/zopefoundation/zope.component/blob/52b524b38efe2d449162788fe8445e948246fd21/src/zope/component/hooks.py#L47-L57). `SiteInfo` is **thread local**. Since the `WSGIServer` layer starts its threads on layer setUp, and only tears them down at layer teardown, its threads start off with a `SiteInfo` instance after initial layer setup that is then detached from the main thread's `SiteInfo` (becomes thread-local) for the lifecycle of the layer.

That would then render `plone.testing.zca`'s attempts to provide component registry isolation via its stacked configuration contexts ineffective, because its `pushGlobalRegistry()` / `popGlobalRegistry()` functions operate on the main thread's site manager, not on the waitress thread one's.

In an attempt to confirm this hypothesis, I changed `SiteInfo` to inherit from `object` instead of `threading.local` - and indeed, this did "fix" the test failure. I was also able to reproduce the test failure with other robot tests, not just `p.a.widgets` - the robot tests from `p.a.multilingual` or `p.a.contenttypes` cause exactly the same issue (if they happen to be run *before* the `plone.restapi` test).

/cc @davisagli

**Comment by Rotonen** [comment]
@tisto I'm trying to get to a starting point for adding daily and weekly test runs to shake the testing stack harder to start finding those out. Additionally running each layer in parallel already passes and that's a pretty decent starting point, as the layer setups happen in isolation with both the `zope.testrunner` `-j` flag and the hacky `mtest` wrapper I'm toying around with.

https://jenkins.plone.org/view/Experimental/job/exp-parallel-tests/40/console
https://github.com/plone/buildout.coredev/blob/roto-testrunner-wrapper-52-py3/bin/mtest3

On the short to medium term I'm planning on doing something like:
1) Do daily sequential `--shuffle` runs
2) Do daily parallel `--shuffle` runs
3) Do weekly sequential 'all possible package pairs` runs
4) Do weekly sequential 'all possible layer pairs` runs
5) Do monthly 'run every testclass alone' runs
6) Do monthly 'run every test alone' runs

Don't worry, we'll get there.

Currently we're still dealing with basic stuff like hitting ulimit as the Robot test setups leak the FD of the socket, though.

**Comment by tisto** [comment]
@Rotonen that all sounds fantastic! Thank you for taking this over!

**Comment by gforcada** [comment]
Regarding single package isolation, we can create a Jenkins job thatvruns only the package tests on every push on that package for each buildout.coredev branch, it should be trivial to wire that up with mr.roboto (tm)

**Comment by Rotonen** [comment]
@gforcada can we do those as pipelines, which share the buildout filesystem? If so, that'd sound like the way to parallelise (and we could get rid of the weight plugin that way too). Don't know if the default should be to split the runs per package or per layer (or both).

But that's something for later, as we'd need to figure out the basics of running our test reliably in any manner.

@tisto thank you for your neat and simple Robot tests, they clued me into `Suite setup` vs. `Test setup`.

https://github.com/kitconcept/robotframework-react/blob/master/tests/create-react-app/test.robot

Does not help to trivially make the Plone tests quicker, though, as most of our Robot suites are very short and we'll close the browser latest at the end of a Suite with how we currently run the tests and spawn the selenium server on demand per suite.

This can be split into two more keywords for suite setup and suite teardown:
https://github.com/plone/plone.app.robotframework/blob/f84d1a90851d69c09b362885ba7c18ed417c84ee/src/plone/app/robotframework/selenium.robot#L44-L51

And then every suite definition should amend to run those new keywords for suite setup and suite teardown.

https://github.com/plone/plone.app.robotframework/blob/f84d1a90851d69c09b362885ba7c18ed417c84ee/src/plone/app/robotframework/tests/test_autologin_library.robot#L9-L10

And we'll need to rearchitect and redesign and reprogram all the Robot tests to have very long suites instead of many short suites, if we want speed.

Or we go with some other technology than Robot. There's only about 100 end to end browser tests in total and the effort level of making them sane to run is about the same that way too.

This ticket has become  a repository dump of miscellaneous testing trivia at this point.

**Comment by gforcada** [comment]
Let's move that to Jenkins.plone.org repo then

**Comment by pbauer** [comment]
All tasks in preparation of 5.2a1 are now done, so I'll close this issue. 
The problem with the thread-local SiteInfo that was unearthed by @lukasgraf in 
https://github.com/plone/Products.CMFPlone/issues/2592#issuecomment-429538716 should maybe become a separate ticket. @lukasgraf: Could you add that please?

**Comment by lukasgraf** [comment]
@pbauer done: plone/plone.testing#58
